datafusion/src/execution/memory_management/mod.rs

```diff
@@ -0,0 +1,405 @@
+                debug!(
+                    "Partition {} released {} from consumer itself {}",
+                    self.partition_id,
+                    released,
+                    consumer.str_repr()
+                );
+                got += memory_manager
+                    .acquire_exec_pool_memory(required - got, consumer_id)
+                    .await;
+            }
+        }
+
+        if got < required {
+            return Err(OutOfMemory(format!(
+                "Unable to acquire {} bytes of memory, got {}",
+                required, got
+            )));
+        }
+
+        debug!("{} acquired {}", consumer_id, got);
+        Ok(got)
+    }
+
+    /// log current memory usage for all consumers in this partition
+    pub async fn show_memory_usage(&self) -> Result<()> {
+        info!("Memory usage for partition {}", self.partition_id);
+        let consumers = self.consumers.lock().await;
+        let mut used = 0;
+        for (_, c) in consumers.iter() {
+            let cur_used = c.get_used();
+            used += cur_used;
+            if cur_used > 0 {
+                info!(
+                    "Consumer {} acquired {}",
+                    c.str_repr(),
+                    human_readable_size(cur_used as usize)
+                )
+            }
+        }
+        let no_consumer_size = self
+            .memory_manager
+            .upgrade()
+            .ok_or_else(|| {
+                DataFusionError::Execution("Failed to get MemoryManager".to_string())
+            })?
+            .exec_memory_used_for_partition(self.partition_id)
+            - (used as usize);
+        info!(
+            "{} bytes of memory were used for partition {} without specific consumer",
+            human_readable_size(no_consumer_size),
+            self.partition_id
+        );
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+/// Id that uniquely identifies a Memory Consumer
+pub struct MemoryConsumerId {
+    /// partition the consumer belongs to
+    pub partition_id: usize,
+    /// unique id
+    pub id: usize,
+}
+
+impl MemoryConsumerId {
+    /// Auto incremented new Id
+    pub fn new(partition_id: usize) -> Self {
+        let id = next_id();
+        Self { partition_id, id }
+    }
+}
+
+impl Display for MemoryConsumerId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.partition_id, self.id)
+    }
+}
+
+#[async_trait]
+/// A memory consumer that supports spilling.
+pub trait MemoryConsumer: Send + Sync + Debug {
+    /// Display name of the consumer
+    fn name(&self) -> String;
+
+    /// Unique id of the consumer
+    fn id(&self) -> &MemoryConsumerId;
+
+    /// Ptr to MemoryManager
+    fn memory_manager(&self) -> Arc<MemoryManager>;
+
+    /// partition that the consumer belongs to
+    fn partition_id(&self) -> usize {
+        self.id().partition_id
+    }
+
+    /// Try allocate `required` bytes as needed
```

Comment:
As I mentioned in the design doc, the distinction between "allocate `required` bytes that I need to operate" and "give me a budget of extra memory I can use to work within" is an important one

External sort can still produce an answer with a small amount of memory, but will potentially go faster with a larger amount.

Reply:
As explained in the previous sort implementation methods above, the memory consumers should try to allocate another bunch of memory in order to continue in-memory computation, or if no more memory is available, this allocate will trigger the partition manager (or just memory-manager if we combine the two) to find out suitable memory consumer(s) and trigger the spills.

So the suggestion here is to adjust the doc or of different design suggested?

Reply:
I think after thinking about this some more (see comment above), I don't have a specific suggestion. Sorry for confusion